Refactor conversion of PHPDoc to type declarations

julienfalque opened the pull request:
This PR:
* refactors parts of `PhpdocToParamTypeFixer` and `PhpdocToReturnTypeFixer` to ease sharing common logic.
* implements some points of #4511

kubawerlos commented:
Could we instead of abstract fixers introduce - as @SpacePossum suggested one day - a namespace of manipulators?

julienfalque commented:
Not sure what you mean exacly?

kubawerlos commented:
Similarly like we have `Tokenizer` we can create `Manipulator` and store there classes to be called by `PhpdocToParamTypeFixer` and `PhpdocToReturnTypeFixer`.

`AbstractPhpdocToTypeDeclarationFixer` ~does not have any property~ has properties that all are in practice constants, thus the class is set of functions that do stuff for the fixers, similar as all [analyzers](https://github.com/FriendsOfPHP/PHP-CS-Fixer/tree/v2.15.3/src/Tokenizer/Analyzer).

julienfalque commented:
Indeed, some parts of the new abstract class could probably be extracted to some "analyzer" class, but some are quite specific to these fixers. I'll have a closer look shortly.

keradus commented:
as this PR is also fixing a bugfix from 2.16 line, shouldn't it actually target 2.16.x as well ?

julienfalque commented:
This PR also introduces new public API in non-internal code (`DocBlock`) so it should be released in a new minor version according to SemVer. Though these new API are more of an detail of this refactoring than an actual new feature, so I'm fine merging this into 2.16. What do you prefer?

keradus commented:
can we backport bugfix only to 2.15?

julienfalque commented:
Well, at the time I created the PR, 2.16 was not released yet so targeting branch master was fine.

The initial purpose of this PR was to refactor things to ease maintenance. While working on it, I noticed it made implementing parts of #4511 easy. When I noticed it also fixes #4605, the work on this PR was already finished. I only added the test cases, which passed without further work. I'm not sure fixing those bugs without leveraging this refactoring is easy. I'll give it a try.

julienfalque commented:
Tests fixed in #4605.

GrahamCampbell commented:
This PR looks good, even if the 2nd follow-up PR to add a new fixer is not merged.

GrahamCampbell commented:
There are merge conflicts. This needs to be rebased, again. 😿

GrahamCampbell commented:
In fact, it looks like https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5092 kinda broke some of the sharing..

GrahamCampbell commented:
Though I guess we can regard adding the config stuff back in as a new feature.

GrahamCampbell commented:
// cc @SpacePossum

GrahamCampbell commented on src/DocBlock/Annotation.php:
```diff
@@ -185,6 +166,33 @@ public function getTag()
+    /**
+     * @return TypeExpression
+     *
+     * @internal until 2.17
```
I don't think we need to do this. Just write internal, and we can delete this tag in 2.17 if we want, or indeed not!

julienfalque replied:
I prefer keeping the version so that if we forget to remove the tag, one still knows this is not internal anymore on 2.17 and higher.

GrahamCampbell commented:
NB, we will also need to add an example showing the scalar types config in use (for the tests to pass when merged up to 2.16). Something like:

```
                new VersionSpecificCodeSample(
                    '<?php
/** @param string|null $bar */
function my_foo($bar)
',
                    new VersionSpecification(70100),
                    ['scalar_types' => false]
                ),
```

julienfalque commented:
Rebased.

This PR now reverts removal of option `scalar_types` from #5092. I don't think the option is a much useful feature but the actual issue was that it was not used. Since this is a documented option, if we want to change/remove it, we should deprecate it first.

To avoid having a release that removes the option then a subsequent one that introduces it again, I suggest we include this in the upcoming 2.16.x release.

> we will also need to add an example showing the scalar types config in use

Added.

GrahamCampbell commented:
Ping @SpacePossum. Can this be merged now?

GrahamCampbell commented:
Looks like there are merge conflicts now. // cc @julienfalque

GrahamCampbell commented:
:+1:

GrahamCampbell commented:
Gah, conflicts again. NB This PR looks great to me @SpacePossum. Would be good to get this merged to avoid repeated merge conflicts.

GrahamCampbell commented:
Oh, this PR also needs updating to support the mixed type. // cc @SpacePossum

GrahamCampbell commented:
Would something like this do:

<img width="668" alt="image" src="https://user-images.githubusercontent.com/2829600/99889173-6cd48080-2c4a-11eb-870f-dfa9c86744ed.png">

julienfalque commented:
I rebased this PR to fix the conflict but did not apply any extra change. Let's merge this first and work on PHP 8 compatibility separately please.

GrahamCampbell commented:
The reason I suggest that change is because support for mixed was added a few days ago, so this PR should not now break it.

GrahamCampbell commented:
I think the php 8 tests now fail after this pr, but were passing before.

GrahamCampbell reviewed (changes requested):
Broken unit tests on php 8

julienfalque commented:
@GrahamCampbell My bad, I didn't notice this was already supported on upstream branch, fixed as you suggested.

GrahamCampbell commented:
Needs rebasing again, against 2.17.

julienfalque commented:
Actually, if I remember correctly, all "new features" are marked `@internal until 2.18` (should it be 3.0?) so it can still be merged into 2.16.

GrahamCampbell commented:
I think 2.17 is the right place for this PR.

julienfalque commented:
Why?

GrahamCampbell commented:
Because it adds new features, and 2.16 should only be getting bug fixes?

julienfalque commented:
The PR:
* adds `scalar_types` option to `phpdoc_to_param_type` rule, but actually this is a revert of #5092 which can be considered a BC break
* adds new APIs that are marked `@internal` for now
* improves an existing fixer, which is something we use to merge in lower branches.

So IMO everything can be merged in 2.16. What do you think @SpacePossum @keradus @kubawerlos?

GrahamCampbell commented on src/DocBlock/Annotation.php:
```diff
@@ -187,6 +168,33 @@ public function getTag()
+    /**
+     * @return TypeExpression
+     *
+     * @internal until 2.18
```
2.19

julienfalque replied:
@keradus Do you plan a 2.19 release or shall we write 3.0 here?

julienfalque replied:
I assumed 2.19 for now.

GrahamCampbell commented on src/DocBlock/Annotation.php:
```diff
@@ -187,6 +168,33 @@ public function getTag()
+    /**
+     * @return null|string
+     *
+     * @internal until 2.18
```
2.19

GrahamCampbell reviewed (changes requested):
Have been testing this PR for a number of weeks, we no issues. Left just a couple of tiny comments.

keradus commented on src/DocBlock/Annotation.php:
```diff
@@ -187,6 +168,33 @@ public function getTag()
+    /**
+     * @return TypeExpression
+     *
+     * @internal until 2.19
```
```suggestion
     * @internal
```

to speed up review. let's separately discuss if we would like to make it public

keradus commented on src/DocBlock/Annotation.php:
```diff
@@ -187,6 +168,33 @@ public function getTag()
+    /**
+     * @return null|string
+     *
+     * @internal until 2.19
```
```suggestion
     * @internal
```

to speed up review. let's separately discuss if we would like to make it applied for BC promise

keradus reviewed (approved):
crazy amount of work. kudos!

keradus commented:
Thank you @julienfalque.
